### ISSUE TITLE
VTT Type Proposal for time-coded general-flash metadata 

### ISSUE THREAD

**Issue by cookiecrook** [w3c/webvtt#512]
Earlier today, Apple released a [new strobing mitigation media accessibility accommodation in tvOS 16.4, iOS 16.4, and macOS 13.3, called "Dim Flashing Lights."](https://developer.apple.com/accessibility/#whats-new) (Additional links below)

Prior to this, the current typical behavior with flashing media is that a viewer is warned *once* before before a video plays. Even in a feature length film, there is no indication of when the flashing may occur or information about the intensity and type of strobing.

On Apple devices, we can now adjust for some strobing/flashing in real time with onboard hardware and software, but as Apple's TV+ content is viewable on many others platforms, we'd also like to be able to do this on non-Apple hardware. 

We think proposing this as an accessibility standard makes sense, as other platforms may wish to adopt their own form of strobing mitigation or warnings.

So we'd like to propose using a WebVTT metadata track to allow streaming services and content owners to denote areas of flashing in video time codes similar to a caption track.

### Minimum proposal
```webvtt
WEBVTT

1
00:00:10.123 --> 00:00:15.432
{
  type: "video.strobing.general-flash",
  level: "75",
}

2
00:00:47.462 --> 00:01:04.028
{
  type: "video.strobing.general-flash",
  level: "100",
}

```

We're not particular to the proposed key names or value formats, but the type value above aligns with an expected update to HLS, and accounts for future expansion into other common strobing patterns such as `red-flash` and `spatial-pattern`.

### Variant proposal that accounts for algorithm and version

Ideally, in order to ensure all content publishers, software developers, and users have the same understanding and representation of the level, the VTT cue format could account for different, documented algorithms and versions for each of those. 

[Apple published a new general flash algorithm (PDF)](https://developer.apple.com/accessibility/downloads/Video-Flashing-Reduction-12.pdf) for use alongside [the related native platform API](https://developer.apple.com/documentation/mediaaccessibility/flashing_lights/responding_to_changes_in_the_flashing_lights_setting) or for [use in other contexts](https://github.com/apple/VideoFlashingReduction/). 

In order to differentiate from other flash pattern algorithms in WebVTT (Harding, for example), the metadata cue could contain a published version and name or URI, such as the following:

```webvtt
WEBVTT

1
00:00:10.123 --> 00:00:15.432
{
  type: "video.strobing.general-flash",
  level: "75",
  test-uri: "https://developer.apple.com/accessibilty/...",
  test-version: "1.0",
}

2
00:00:47.462 --> 00:01:04.028
{
  type: "video.strobing.general-flash",
  level: "100",
  test-uri: "https://developer.apple.com/accessibilty/...",
  test-version: "1.0",
}

```

We're not particular to the proposed key names or values, and look forward to discussing further with the Working Group. 

### Prerequisite:
I have filed another issue we believe is important to consider alongside this one.

- #511

### Other documentation on the recently released feature

- [Dim Flashing Lights Overview](https://developer.apple.com/accessibility/#whats-new)
- [Explainer and Algorithms (PDF)](https://developer.apple.com/accessibility/downloads/Video-Flashing-Reduction-12.pdf)
- [Sample Code: Responding to changes in the flashing lights setting](https://developer.apple.com/documentation/mediaaccessibility/flashing_lights/responding_to_changes_in_the_flashing_lights_setting)
- [GitHub Sample Code: Detection of flashing lights in video content](https://github.com/apple/VideoFlashingReduction/)

**Comment by cookiecrook** [comment]
@eric-carlson @chrisn

**Comment by gkatsev** [comment]
Dim Flashing Lights looks fascinating!

This is an interesting proposal. I think this would be the first time someone asked to standardize a specific metadata format in WebVTT. I think historically, this has probably been left to users of WebVTT. Furthermore, I don't think we'd want it in WebVTT itself, but could still be a related spec or something. Not exactly sure what the options available are. Happy to discuss it further.

Also, I'm wondering how this would be represented in HLS. Would it allow for segmented VTT and then the main HLS playlist would reflect that his is a Dim Flashing Lights metadata playlist?

**Comment by cookiecrook** [comment]
I'll ping some of the HLS folks to reply to those questions.

**Comment by chrisn** [comment]
I have a couple more questions, following discussion in today's [TTWG meeting](https://www.w3.org/2023/03/30-tt-minutes.html):
* In your view, does this use case motivate the use of a metadata-specific cue API, i.e., [DataCue](https://github.com/WICG/datacue/blob/main/explainer.md), rather than using VTTCue?
* What are the synchronization requirements for handling of the time codes. For example, is frame accuracy needed? There's a long discussion in https://github.com/w3c/media-and-entertainment/issues/4 on achieving frame accurate synchronization with the HTML `<video>` element.

**Comment by eric-carlson** [comment]
> I have a couple more questions, following discussion in today's [TTWG meeting](https://www.w3.org/2023/03/30-tt-minutes.html#t05):
> 
> * In your view, does this use case motivate the use of a metadata-specific cue API, i.e., [DataCue](https://github.com/WICG/datacue/blob/main/explainer.md), rather than using VTTCue?

Are you suggesting that all`metadata` tracks should emit [DataCue](https://github.com/WICG/datacue/blob/main/explainer.md)? 

If so, what about do we about existing `metadata` formats like [WebVMT](https://www.w3.org/2021/sdw/proposals/geotagging/webvmt/).

If not, what makes this JSON-based metadata cue format more suitable for DataCue?

> * What are the synchronization requirements for handling of the time codes? For example, is frame accuracy needed? There's a long discussion in [Frame accurate seeking of HTML5 MediaElement media-and-entertainment#4](https://github.com/w3c/media-and-entertainment/issues/4) on achieving frame accurate synchronization with the HTML `<video>` element.

While it is important for the UA or a script to be able to process general-flash metadata in time to dim the video to block the strobing/flashing, I don't think "frame accuracy" is needed (I don't believe it is possible to guarantee frame accurate delivery of JavaScript events in any case).

**Comment by chrisn** [comment]
> Are you suggesting that all `metadata` tracks should emit DataCue?

No, this would be a breaking change, although in principle a new `kind` value could be used to request DataCues.

> If not, what makes this JSON-based metadata cue format more suitable for DataCue?

In this (and the WebVMT use case), the main benefit seems to be a saving a `JSON.parse()` call, which wouldn't be a huge improvement.

I was looking in general for motivation to progress DataCue. But given VTTCue can work well enough in these cases I'm not seeing much reason progress it.

> While it is important for the UA or a script to be able to process general-flash metadata in time to dim the video to block the strobing/flashing, I don't think "frame accuracy" is needed (I don't believe it is possible to guarantee frame accurate delivery of JavaScript events in any case).

I agree, and so authors or generators of the metadata need to take into account typical client processing time.

**Comment by gkatsev** [comment]
> > If not, what makes this JSON-based metadata cue format more suitable for DataCue?
> 
> In this (and the WebVMT use case), the main benefit seems to be a saving a `JSON.parse()` call, which wouldn't be a huge improvement.

Potentially, can extend VTTCue to have a `.json` property that returns the cue text parsed as json.

**Comment by nigelmegitt** [comment]
Seems to me a bit the wrong way around to extend VTTCue where the real use case is removing unnecessary stuff, plus adding maybe one thing.

It would make architectural sense to support standardisation of metadata formats outside of WebVTT, and to have something like a `DataCue` that offers the same timing functionality as `VTTCue` and keeps a `.text` attribute for compatibility, but doesn't bother with the presumably completely unused methods and attributes for presenting text. I'm not sure if anything else would actually be needed to avoid this being a breaking change - possibly if people directly test that the cue object is `VTTCue` (not sure why anyone would)? 

The benefits could be:
a) Allows the UA to parse the JSON in its own time, or when needed, whichever is the sooner, thus being more likely able to meet the timing requests for `onenter()` and `onexit()` handing.
b) Saves that `json.parse()` call.
c) Saves a small amount of per-cue-object memory associated with the pointers to the unused and not needed `VTTCue` attributes and methods associated with cue text presentation.

**Comment by nigelmegitt** [comment]
If tying a `DataCue` to JSON is unhelpful, because it isn't always needed, perhaps there may at some point in the future be a useful subclass of `DataCue` optimised for JSON processing, a `JSONCue` for example.

**Comment by cookiecrook** [comment]
I plan to attend that day, and will try to bring others to the meeting as well. Thanks.

**Comment by nigelmegitt** [comment]
This is on the agenda for the call scheduled for [April 27 15:00 UTC](https://www.w3.org/events/meetings/fb1d4436-271d-4b22-812c-986f9672d8f0/20230427T150000).

**Comment by cookiecrook** [comment]
Minor editorial corrections in the [official meeting minutes](https://www.w3.org/2023/04/27-tt-minutes.html).

**Comment by cookiecrook** [comment]
I realized after the call we had a vocabulary mismatch due to my unfamiliarity with WebVMT. The term "mapping data" came up a few times, which I and at least one more person took as the general meaning: "associating data." It's now clear in the context of WebVMT that others meant literal **_cartography_** data. 😉

After the fact, it's clear WebVMT was referred to as an example only, not a proposed format. Hopefully this acknowledgement clears up some of the communication breakdown regarding whether this should be a Rec, Note, or Incubation referring to some other normative Rec.

**Comment by cookiecrook** [comment]
Follow-up on the HLS Docs:
https://developer.apple.com/documentation/http-live-streaming/hls-authoring-specification-for-apple-devices-appendixes#Dim-Flashing-Lights

**Comment by japuiglaucia** [comment]
VTT Type Proposal for time-coded general-flash metadata #512

**Comment by cookiecrook** [comment]
Status: Still blocked, but getting warmer. 

The older ATTRIBUTES block approach was abandoned last month in favor of this new prerequisite:
- #548
